### PR TITLE
v1.1 -> v2.0 Semantics refactor

### PR DESCRIPTION
**Pull request by ejhumphrey**
Renames tests for clearer mappings between submodules
Addresses much of the conversation from #75 
- will break the current API, and needs some help signaling deprecations
- `Streamer.generate` is now `Streamer.iterate`, since it returns an iterator
- `Streamer`types are now directly iterable
- `Mux` semantics are simplified to remove the (now superfluous) `seed` concept, which were synonymous with streams; now there are only streamers and streams.
- the notion of `batch` is scrubbed in all core functionality.

Will resolve: #64, #75, #76, #77, #78, #79, #81

### CODE REVIEW

File: pescador/buffered.py [hunk 1]
```diff
@@ -56,31 +58,31 @@ def activate(self):
         """Activates the stream."""
         self.stream_ = self.streamer
 
-    def generate(self, max_batches=None):
+    def iterate(self, max_iter=None):
```
Comment (bmcfee) [inline pescador/buffered.py]: persist the old generate function here.  we can wrap the call signature for API compatibility, but there should be a warning.

File: pescador/buffered.py [hunk 2]
```diff
@@ -56,31 +58,31 @@ def activate(self):
...
-                if max_batches is not None and n >= max_batches:
+                if max_iter is not None and n >= max_iter:
                     break
                 yield batch
 
 
-def buffer_batch(generator, buffer_size):
-    '''Buffer an iterable of batches into larger (or smaller) batches
+def buffer_batch(streamer, buffer_size):
+    '''Buffer data samples from an streamer into one data object.
```
Comment (bmcfee) [inline pescador/buffered.py]: this doesn't actually require a streamer, right?

File: pescador/core.py [hunk 3]
```diff
@@ -8,6 +8,11 @@
 
 class StreamActivator(object):
     def __init__(self, streamer):
+        for mname in ['activate', 'deactivate']:
```
Comment (bmcfee) [inline pescador/core.py]: why not just check `isinstance(streamer, Streamer)` instead?

File: pescador/core.py [hunk 4]
```diff
@@ -92,8 +98,8 @@ def __init__(self, streamer, *args, **kwargs):
 
         if not (inspect.isgeneratorfunction(streamer) or
                 isinstance(streamer, (collections.Iterable, Streamer))):
-            raise PescadorError('streamer must be a generator, iterable, or '
-                                'Streamer')
+            raise PescadorError('`streamer` must be a callable function that '
+                                'returns an iterable object.')
```
Comment (bmcfee) [inline pescador/core.py]: could we actually generalize this and support iterables as well as functions?  Call it a stretch goal.

File: pescador/core.py [hunk 5]
```diff
@@ -113,57 +119,53 @@ def activate(self):
...
             # If it's iterable, use it directly.
-            self.stream_ = self.streamer
+            self.stream_ = iter(self.streamer)
 
     def deactivate(self):
         self.stream_ = None
 
-    def generate(self, max_batches=None):
-        '''Instantiate the generator
+    def iterate(self, max_iter=None):
```
Comment (bmcfee) [inline pescador/core.py]: again, deprecation rename / persist old signature with warning

File: pescador/core.py [hunk 6]
```diff
@@ -113,57 +119,53 @@ def activate(self):
...
         while True:
-            for item in self.generate():
-                yield item
+            for obj in self:
+                yield obj
 
+    # TODO: ejhumphrey would like to deprecate this class method and turn it
+    #       into a "transform". @bmcfee please advise? example to follow?
     def tuples(self, *items, **kwargs):
         '''Generate data in tuple-form instead of dicts.
```
Comment (bmcfee) [inline pescador/core.py]: I'm still not 100% on what a `transform` is defined to be.

I find the `.tuples()` interface quite useful for interfacing with keras though, and I'd prefer to not have to replace it with something more cumbersome.

File: pescador/mux.py [hunk 7]
```diff
@@ -19,24 +19,26 @@ class Mux(core.Streamer):
...
     ...     MY_FUNCTION(batch)
+
+    Mux([stream, range(8), stream2])
     '''
 
-    def __init__(self, seed_pool, k,
-                 lam=256.0, pool_weights=None, with_replacement=True,
-                 prune_empty_seeds=True, revive=False,
+    def __init__(self, streamers, k,
+                 lam=256.0, weights=None, with_replacement=True,
```
Comment (bmcfee) [inline pescador/mux.py]: deprecation renames EVERYWHERE

while we're at it though, I'd like to replace `lam`=>`rate`

File: pescador/zmq_stream.py [hunk 8]
```diff
@@ -81,15 +81,16 @@ def zmq_recv_batch(socket, flags=0, copy=True, track=False):
     return results
 
 
-def zmq_worker(port, streamer, terminate, copy=False, max_batches=None):
+def zmq_worker(port, streamer, terminate, copy=False, max_iter=None):
```
Comment (bmcfee) [inline pescador/zmq_stream.py]: renames

**Comment by bmcfee** [comment]
@ejhumphrey Re: handling variable rename deprecations, here's how librosa does it.  Note that the current stable version has no lingering renames, but rolling back to 0.4.3 you can find some examples.

1. Check out the `util.deprecation` [submodule](https://github.com/librosa/librosa/blob/446708889bf04438aaac7fe64aa738a09955bbf6/librosa/util/deprecation.py)
2. Here's an example of it in use: https://github.com/librosa/librosa/blob/446708889bf04438aaac7fe64aa738a09955bbf6/librosa/core/constantq.py#L327-L329
3. There's also some helpers in the `util.decorators` [submodule](https://github.com/librosa/librosa/blob/446708889bf04438aaac7fe64aa738a09955bbf6/librosa/util/decorators.py) for deprecating entire functions, or moving them to different submodules/renaming them.
4. Here's an example of move/rename behavior: https://github.com/librosa/librosa/blob/446708889bf04438aaac7fe64aa738a09955bbf6/librosa/core/audio.py#L692-L693
5. Here's an example of deprecated function behavior: https://github.com/librosa/librosa/blob/446708889bf04438aaac7fe64aa738a09955bbf6/librosa/feature/spectral.py#L1337-L1338

The [futurepast](https://github.com/amueller/futurepast) package implements some of these things and more, but it's not yet at an initial release.

File: pescador/buffered.py (hunk 1 above)
Comment (ejhumphrey) [inline pescador/buffered.py]: roger that

File: pescador/buffered.py (hunk 2 above)
Comment (ejhumphrey) [inline pescador/buffered.py]: no, it needs an iterable; not sure what a sane variable name would be here. I also aspire to change this on a subsequent PR, so .... any strong opinions?

File: pescador/core.py (hunk 3 above)
Comment (ejhumphrey) [inline pescador/core.py]: **tl;dr:** ya, that's better.

**commentary:** I think I had a lapse of Java ptsd, and wanted to check that the object implements the interface of `activate`/`deactivate`, since it doesn't strictly matter that it's a Streamer (although when would we run into that? /shrug). If we really cared, a more Pythonic way of doing this would be to define an ActiveStream mixin, but ... this is total overkill for where we are right now, and mention it now only for completeness.

File: pescador/core.py (hunk 4 above)
Comment (ejhumphrey) [inline pescador/core.py]: Actually it does as of [this change](https://github.com/pescadores/pescador/pull/80/files/44d82c225dd8ceb6777cead497a7682ee4eb195c#diff-535170d9182a6f744ecc1f98d69624d5R124) 😄  ... I've updated the error message, and added a `Streamer(iterable)` test in da42dbf1.

File: pescador/core.py (hunk 5 above)
Comment (ejhumphrey) [inline pescador/core.py]: roger roger

File: pescador/core.py (hunk 6 above)
Comment (ejhumphrey) [inline pescador/core.py]: I've added a [long-form response here](https://github.com/pescadores/pescador/issues/77#issuecomment-292806012), including why I think this is an anti-pattern.

**tl;dr:** `Streamer.tuples` only works for a subset of data streams that might be fed through it; I'd propose explicitly separating `Streamer` functionality from processing data streams. 

The thing I'd like to replace it with is a separate `tuples` function (as sketched out in that linked issue), which I'd argue isn't any more cumbersome than what we currently have and Streamer remains data stream agnostic.

File: pescador/mux.py (hunk 7 above)
Comment (ejhumphrey) [inline pescador/mux.py]: haha roger

File: pescador/zmq_stream.py (hunk 8 above)
Comment (ejhumphrey) [inline pescador/zmq_stream.py]: yayaya

File: pescador/buffered.py (hunk 1 above)
Comment (ejhumphrey) [inline pescador/buffered.py]: @bmcfee do you have an example of "@move"ing a class's method? I could hack it out of if need be, but I'd rather do the fast thing.

File: pescador/buffered.py (hunk 2 above)
Comment (ejhumphrey) [inline pescador/buffered.py]: I've opted for `stream`

**Comment by ejhumphrey** [comment]
thanks @bmcfee, made it pretty far ... still have some work to do, but any advice / pointers will be helpful.

**Comment by bmcfee** [comment]
@ejhumphrey I don't have an example of `@move`ing a class method, because that gets a bit more involved. However, I think you don't necessarily want or need that here, since the call signature has changed as well.  I would just write a shim function to translate to the new API, and a custom warning message.

**Comment by cjacoby** [comment]
(I have made a first pass looking through this; I won't have time for another look until tomorrow, because work+taxes)

**Comment by ejhumphrey** [comment]
guh, I need to get the test coverage back up (idk what's up with that), but this is in a state where it's worth taking another look @bmcfee ... I think I did most of the deprecation stuff correctly, but do let me know if that's not the case (I have no doubt you will :o)

File: pescador/maps.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    y : np.ndarray, list of np.ndarray, or None
+        If `outputs` is a string, `y` is a single np.ndarray.
+        If `outputs` is an iterable of strings, `y` is a list of np.ndarrays.
+        If `outputs` is a null type, `y` is None.
+    """
+    if inputs and isinstance(inputs, six.string_types):
+        inputs = [inputs]
+    if outputs and isinstance(outputs, six.string_types):
+        outputs = [outputs]
+
```
Comment (bmcfee) [inline pescador/maps.py]: I don't think we should upcast to list here.  Keras's interface is overloaded to support tensors or lists of tensors as inputs (or outputs). The following are all perfectly valid:

```python
model = K.models.Model(x1, y1)
model = K.models.Model(x1, [y1, y2])
model = K.models.Model([x1, x2], [y1, y2])
```
where `x*` and `y*` are tensors.

`keras_tuples` should let me do any of the following:

```python
keras_tuples(streamer, inputs='X1', outputs='Y1')  # generates tuples of (ndarray, ndarray)
keras_tuples(streamer, inputs='X1', outputs=['Y1', 'Y2'])  # generates tuples of (ndarray, list of ndarray)
keras_tuples(streamer, inputs=['X1'], outputs=['Y1', 'Y2'])  # generates tuples of (list of ndarray, list of ndarray)
```

File: pescador/zmq_stream.py [hunk 10]
```diff
@@ -181,18 +180,19 @@ def generate(self, max_batches=None):
...
 
             # Yield from the queue as long as it's open
             while True:
-                yield zmq_recv_batch(socket)
+                yield zmq_recv_data(socket)
 
         except StopIteration:
             pass
 
+        # TODO: Fix bare except
```
Comment (bmcfee) [inline pescador/zmq_stream.py]: no, this is an intentional bare except that gets reraised.  IIRC, this is here because of some weirdness with transportable exceptions.

File: pescador/maps.py (hunk 9 above)
Comment (ejhumphrey) [inline pescador/maps.py]: cool, this is helpful. I'm not super familiar with the `fit` interface, so I was trying to get as far as possible from the docs.

File: pescador/zmq_stream.py (hunk 10 above)
Comment (ejhumphrey) [inline pescador/zmq_stream.py]: gotcha, I'll nuke the comment ... I was just following my linter blind

File: pescador/maps.py (hunk 9 above)
Comment (bmcfee) [inline pescador/maps.py]: ahahhahahahhaha

Documentation?

File: pescador/zmq_stream.py (hunk 10 above)
Comment (bmcfee) [inline pescador/zmq_stream.py]: feel free to add a pylint disable comment there.

File: pescador/maps.py (hunk 9 above)
Comment (ejhumphrey) [inline pescador/maps.py]: now that I'm reading this on a computer, rather than my phone, I've discovered that I've written the behavior you desire ... although perhaps in not the most elegant way. 

The *keys* get upcast to a list, but singleton items are downcast to np.ndarrays; see L112, L116.

This was the first semi-clean way I saw toward the end of the night to handle both string / list inputs, and return np.ndarray / list outputs.

**Comment by ejhumphrey** [comment]
Alrighty, I think this PR is all buttoned up; tests pass, and fixed the drop in coverage. Just need signoff from y'all (@bmcfee & @cjacoby)

**Comment by cjacoby** [comment]
Apologies; I'm running around like crazy this week in preparation for flying to NYC tomorrow. If i don't get to it today (unlikely, honestly), I'll catch up on all this on the plane tomorrow. 

If that's blocking you and you want to push ahead, either just do it or poke me "offline" and I'll bump it up in my queue.

File: pescador/buffered.py [hunk 11]
```diff
@@ -46,6 +48,12 @@ def __init__(self, streamer, buffer_size,
              completing teh last batch, will just return the number
              of samples currently in the buffer.
         """
+        warn('`BufferedStreamer` is deprecated in 1.1.0 '
+             'This functionality is superseded by the generator function '
+             '`pescador.buffer_data` in 2.0.0, which can be used with '
+             '`pescador.Streamer` to similar ends.'
+             'Use this idiom instead to maintain forwards compatibility.',
+             DeprecationWarning)
```
Comment (bmcfee) [inline pescador/buffered.py]: Are the examples in the documentation up to date with the latest jazz?

File: pescador/core.py [hunk 12]
```diff
@@ -35,54 +41,54 @@ class Streamer(object):
...
+    ...     print(i)  # Displays 0, 1, 2, 3, 4, 0, 1, 2, ...
 
     An alternate interface for the same:
 
-    >>> for i in GS(cycle=True):
-    ...     print(i)
+    >>> for i in stream(cycle=True):
+    ...     print(i)  # Displays 0, 1, 2, 3, 4, 0, 1, 2, ...
+
+    NOTE: We don't use cycle because it's a closure and can't be serialized.
```
Comment (bmcfee) [inline pescador/core.py]: this note shouldn't be in the docstring

File: pescador/core.py [hunk 13]
```diff
@@ -213,40 +225,49 @@ def tuples(self, *items, **kwargs):
...
         See Also
         --------
-        generate
+        iterate
         cycle
-        tuples
         '''
+        max_iter = rename_kw('max_batches', max_batches,
+                             'max_iter', max_iter,
+                             '1.1.0', '2.0.0')
```
Comment (bmcfee) [inline pescador/core.py]: Please be consistent about major/minor version numbers in deprecations.  I see `1.1` and `2.0` elsewhere.

File: pescador/core.py [hunk 14]
```diff
@@ -204,7 +213,10 @@ def tuples(self, *items, **kwargs):
         generate
         cycle
         '''
-
+        warn('`Streamer.tuples()` is deprecated in 1.1.0 '
+             'This functionality is moved to `pescador.tuples` in 2.0.0. '
```
Comment (bmcfee) [inline pescador/core.py]: is there actually a `pescador.tuples`?  I only see `maps.keras_tuples`

File: pescador/maps.py (hunk 9 above)
Comment (bmcfee) [inline pescador/maps.py]: I think we also want an old-style tuples map that works like `Streamer.tuples` does/did:

```python
def tuples(gen, *args):
    for item in gen:
         yield (item[x] for x in args)
```
or something

File: tests/test_core.py [hunk 15]
```diff
@@ -1,23 +1,53 @@
...
 
-import pescador
+import pescador.core
 import test_utils as T
 
 
-def test_streamer_list():
+def test_StreamActivator():
+    with pytest.raises(pescador.core.PescadorError):
+        pescador.core.StreamActivator(range)
```
Comment (bmcfee) [inline tests/test_core.py]: this only tests the failure mode?  I guess that's okay.

File: tests/test_maps.py [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+            for n in range(10)]
+
+    stream = pescador.maps.keras_tuples(data, inputs="foo", outputs="bar")
+    for n, (x, y) in enumerate(stream):
+        assert n == x == y * 2
+
+    stream = pescador.maps.keras_tuples(data, outputs="whiz")
+    for n, (x, y) in enumerate(stream):
+        assert n == y / 2.
+        assert x is None
```
Comment (bmcfee) [inline tests/test_maps.py]: Should this be` None` or `[]`?

File: pescador/buffered.py (hunk 11 above)
Comment (ejhumphrey) [inline pescador/buffered.py]: no, probably not. I'll look into that now

File: pescador/core.py (hunk 12 above)
Comment (ejhumphrey) [inline pescador/core.py]: ha, right. that was a note for me to migrate into documentation elsewhere

File: pescador/core.py (hunk 14 above)
Comment (ejhumphrey) [inline pescador/core.py]: nope, that's stale

File: pescador/core.py (hunk 13 above)
Comment (ejhumphrey) [inline pescador/core.py]: is 1.1 fine? or 1.1.0?

File: tests/test_maps.py (hunk 16 above)
Comment (ejhumphrey) [inline tests/test_maps.py]: [keras docs](https://keras.io/models/model/#fit) (currently) say `None`

File: pescador/core.py (hunk 13 above)
Comment (bmcfee) [inline pescador/core.py]: i'd go with 1.1; sub-minors should never be involved in deprecations.

File: examples/frameworks/keras_example.py [hunk 17]
```diff
@@ -93,71 +117,99 @@ def build_model(input_shape):
...
-    epochs=epochs,
-    verbose=1,
-    validation_data=(X_test, Y_test))
-
-score = model.evaluate(X_test, Y_test, verbose=0)
-print('Test score:', score[0])
-print('Test accuracy:', score[1])
+try:
+    model.fit_generator(
+        pescador.tuples(batches, 'X', 'y'),
```
Comment (bmcfee) [inline examples/frameworks/keras_example.py]: `pescador.tuples(batches.cycle(), 'X', 'y')` ?

Comment (ejhumphrey) [inline examples/frameworks/keras_example.py]: batches is infinite anyways?

**Comment by ejhumphrey** [comment]
alrighty, updated the examples, and added some slightly more meaningful benchmarking to the zmq_example, which burns cycles to prove that things are happening in the background.

LET'S DO THIS

File: examples/frameworks/keras_example.py [hunk 18]
```diff
@@ -93,71 +118,101 @@ def build_model(input_shape):
...
-def data_generator(X, y):
-    '''A basic generator for the data.'''
+    Yields
+    ------
+    data : dict
+        Single image sample, like {X: np.ndarray, y: np.ndarray}
+    '''
     X = np.atleast_2d(X)
-    # y's are binary vectors, and should be of shape (1, 10) after this.
+    # y's are binary vectors, and should be of shape (10,) after this.
```
Comment (bmcfee) [inline examples/frameworks/keras_example.py]: atleast_2d will make them shape (1,10)

File: examples/mux_files_example.py [hunk 19]
```diff
@@ -61,14 +61,14 @@ def npz_generator(npz_path):
     """Generate data from an npz file."""
     npz_data = np.load(npz_path)
     X = npz_data['X']
-    # y's are binary vectors, and should be of shape (1, 10) after this.
+    # y's are binary vectors, and should be of shape (10,) after this.
```
Comment (bmcfee) [inline examples/mux_files_example.py]: same comment here, i think?

File: pescador/maps.py (hunk 9 above)
Comment (bmcfee) [inline pescador/maps.py]: Upon closer inspection, I think this still isn't correct.

I don't think `fit(inputs=x)` and `fit(inputs=[x])` do the same thing (I might be wrong on that though).  If that's the case, then upcast-downcast will not behave as expected.

File: pescador/mux.py [hunk 20]
```diff
@@ -83,14 +90,48 @@ def __init__(self, seed_pool, k,
 
             If None, the random number generator is the RandomState instance
             used by np.random.
+
+        seed_pool : iterable of streamers
+            .. warning:: This parameter name was deprecated in pescador 1.1
+            Use the `streamers` parameter instead.
+            The `seed_pool` parameter will be removed in pescador 2.0.
+
+        lam : float > 0.0
```
Comment (cjacoby) [inline pescador/mux.py]: :100: on converting `lam` to `rate` - I think that's sooo much more intuitive.

However, you missed including `rate` in the docs here. I suspect the `lam` above is supposed to be `rate`.

File: pescador/version.py [hunk 21]
```diff
@@ -3,4 +3,4 @@
 """Version info"""
 
 short_version = '1.0'
-version = '1.0.0'
+version = '1.1.0'
```
Comment (cjacoby) [inline pescador/version.py]: @bmcfee - Correct me if I misunderstand the case, but I believe last time I tried to do this that you requested all version increments be in a separate PR. I would expect that to apply here.

Also... shouldn't we increment the `short_version` too?

File: tests/test_buffered.py [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+
+    gen_stream = pescador.Streamer(T.md_generator, dimension, 50,
+                                   size=batch_size)
+    estimate = pescador.BufferedStreamer(gen_stream, buf_size)
+
+    estimate = list(__unpack_stream(estimate))
+
+    T.__eq_lists(reference, estimate)
+
+    estimate = pescador.BufferedStreamer(gen_stream, buf_size)
```
Comment (cjacoby) [inline tests/test_buffered.py]: curious; is this copypasta, or have you kept the old interface here on purpose? Same below.

(If these tests are here to make sure the old interface still works, that's fine, just mention that in a comment someplace, and that they should be removed when the deprecations are removed.)

File: examples/frameworks/keras_example.py (hunk 18 above)
Comment (ejhumphrey) [inline examples/frameworks/keras_example.py]: done

File: examples/mux_files_example.py (hunk 19 above)
Comment (ejhumphrey) [inline examples/mux_files_example.py]: done

File: pescador/mux.py (hunk 20 above)
Comment (ejhumphrey) [inline pescador/mux.py]: yup, caught it

File: pescador/buffered.py [hunk 23]
```diff
@@ -56,151 +63,30 @@ def activate(self):
...
         ----------
-        max_batches : int
-            For the BufferedStreamer, max_batches is the
+        max_iter : int
+            For the BufferedStreamer, max_iter is the
             number of *buffered* batches that are generated,
             not the number of individual samples.
+
+        partial : bool, default=True
+            Return buffers smaller than the requested size.
```
Comment (ejhumphrey) [inline pescador/buffered.py]: done

File: pescador/version.py (hunk 21 above)
Comment (ejhumphrey) [inline pescador/version.py]: can I trouble y'all for consensus? I'd love to get this over the goal line today 😄

File: tests/test_buffered.py (hunk 22 above)
Comment (ejhumphrey) [inline tests/test_buffered.py]: ya, the whole `buffered` submodule is going away in 2.0, I've added a comment.

**Comment by ejhumphrey** [comment]
couple comments here because this thread is getting *insane*.
- versioning: what's our stance (this vs separate)?
- docs: help would be great. also I dunno what I'm doing. so help would be *really* great.

File: pescador/version.py (hunk 21 above)
Comment (cjacoby) [inline pescador/version.py]: Plz see: https://github.com/pescadores/pescador/blob/master/CONTRIBUTING.md, "Additional Comments".

**Comment by cjacoby** [comment]
Re: Versioning. I'm posting this here, too 'cause embedded is getting too crazy.

The `CONTRIBUTING.md` says:
```Please keep branches as simple as possible: do not implement multiple features in the same branch, unless they are mutually dependent on one another. By the same token, please do not change unrelated portions of the repository (e.g., version numbers or changelog entries).```
and by that, I'd say, separate PR for version update.

Re: Docs - let's start an Issue with a checklist?

**Comment by cjacoby** [comment]
(Once we have those two things, I think this PR is 👍

File: pescador/version.py (hunk 21 above)
Comment (ejhumphrey) [inline pescador/version.py]: roger roger

**Comment by ejhumphrey** [comment]
hey team, two things: 
 - for clarity, you're saying create a docs issue **and** fix it in this PR? or create an issue and fix it elsewhere?
- in parallel, I started updating a recent project to (what will be) pescador 1.1, and having some difficulties dealing with `Mux`, specifically, the `Mux-of-Muxes` case that remains a little bit suspicious, as far as I'm concerned. I'm still in the process of determining whether my current struggles are a result of poor initial design of data streams in that project, a fundamental issue with the Mux API, or some combination of both. Regardless, I don't want to merge this PR until I've got concrete answers about what the hell is going on.

**Comment by cjacoby** [comment]
* Re: docs. I'm proposing it gets fixed in a separate PR. The items could get logged in that PR, or in an issue. But I'd start with an issue at least. If you write up an issue with all the components that changed such that they will affect the docs, I could take a stab at updating the docs to reflect the changes made in a PR. 

* Re: Muxes. This sounds like a good candidate for identifying via improved testing, ya?

File: pescador/mux.py [hunk 24]
```diff
@@ -19,59 +20,65 @@ class Mux(core.Streamer):
...
-            If ``True``, Streamers that previously produced no data are never
+        prune_empty_streams : bool
+            Disable streamers that produce no data.
+            If ``True``, streamers that previously produced no data are never
             revisited.
             Note that this may be undesireable for streams where past emptiness
             may not imply future emptiness.
 
         revive: bool
             If ``with_replacement`` is ``False``, setting ``revive=True``
```
Comment (ejhumphrey) [inline pescador/mux.py]: @bmcfee @cjacoby, pop quiz ... if `with_replacement=True`, what is the significance of `revive`?

Comment (cjacoby) [inline pescador/mux.py]: Nothing.

**Comment by ejhumphrey** [comment]
@mcfee, can I trouble you for a ✅ ?

File: pescador/mux.py [hunk 25]
```diff
@@ -104,45 +145,48 @@ def __init__(self, seed_pool, k,
...
 
         for idx in range(self.k):
 
-            if not (self.seed_distribution > 0).any():
+            if not (self.distribution > 0).any():
                 break
 
+            # TODO: Stream uniqueness can't be guaranteed here.
+            # Other machinery necessary to impose permutation, this is
+            # implicitly `with_replacement`
```
Comment (bmcfee) [inline pescador/mux.py]: really?  I thought we zero out the probability of active streams when `with_replacement=False`.

Comment (ejhumphrey) [inline pescador/mux.py]: You're correct, and I've already removed the comment locally; this was my read at midnight last night, and fixed it this am with some coffee-inspired clarity.

Comment (ejhumphrey) [inline pescador/mux.py]: btw I was totally wrong and just thought i did (uhderp)

**Comment by ejhumphrey** [comment]
replaced by #88
